### PR TITLE
Add Language Model Filter Builder

### PR DESCRIPTION
**Pull request by andyflury**
# Add Language Model Filter Builder

## Description
This PR introduces a new module that converts natural language queries into LangChain4j Filter objects using language models and structured outputs, enabling hybrid search capabilities for RAG applications.

Users can now write queries like "Find documents by John Doe published after 2023 with rating above 4 about machine learning" and get both a structured filter for metadata constraints and a clean semantic query for embedding search.

## Features
- 🤖 **Natural Language Processing**: Converts user queries into structured filters automatically
- 🎯 **Hybrid Search**: Separates semantic content from metadata constraints  
- 🔧 **Type Safety**: Strongly typed column definitions with automatic type conversion
- 📊 **Comprehensive Filter Support**: All LangChain4j filter types (comparison, set, logical operations)
- 📅 **Smart Date Handling**: Processes relative dates ("last week", "this month")
- ✅ **Reliable Outputs**: Uses JSON Schema-based structured outputs for consistency
- 🔍 **Enum Support**: Automatic handling of enum types with value validation

## Use Cases
- RAG applications requiring metadata filtering alongside semantic search
- Natural language database/document queries
- Content management systems with intuitive search
- Advanced search interfaces for technical documentation
- Customer support systems with contextual filtering

### Architecture
The module uses a multi-step process:
1. **Schema Definition**: `TableDefinition` describes available metadata columns with types and descriptions
2. **Prompt Generation**: Creates structured prompts with column information and current date context
3. **LLM Processing**: Language model generates JSON using ResponseFormat.JSON with schema validation
4. **Filter Parsing**: JSON parsed into LangChain4j Filter objects with automatic type conversion
5. **Query Modification**: Semantic content extracted and cleaned for embedding search

### Key Classes
- `LanguageModelJsonFilterBuilder`: Main conversion class
- `TableDefinition`: Metadata schema definition with builder pattern
- `FilterResult`: Contains generated filter and modified query
- `FilterSchemas`: JSON schemas for structured LLM outputs

## Example Usage

```java
// Define metadata schema
TableDefinition tableDefinition = TableDefinition.builder()
    .addColumn("author", String.class, "Document author")
    .addColumn("publishDate", LocalDate.class, "Publication date")
    .addColumn("rating", Number.class, "User rating from 1-5")
    .addColumn("category", DocumentCategory.class, "Document category")
    .build();

// Create filter builder
LanguageModelJsonFilterBuilder builder = new LanguageModelJsonFilterBuilder(
    chatModel, tableDefinition);

// Convert natural language to filter + semantic query
FilterResult result = builder.buildFilterAndQuery(
    "Find research papers by Alice published after 2023-01-01 with rating above 4 about machine learning");

// Use results for hybrid search
Filter filter = result.getFilter(); 
// -> AND(IsEqualTo("author", "Alice"), AND(IsEqualTo("category", "RESEARCH"), 
//        AND(IsGreaterThan("publishDate", "2023-01-01"), IsGreaterThan("rating", 4))))

String semanticQuery = result.getModifiedQuery(); 
// -> "machine learning"

// Search with both semantic similarity and metadata constraints
List<EmbeddingMatch<TextSegment>> results = embeddingStore.findRelevant(
    embeddingModel.embed(semanticQuery).content(), 
    10, 0.7, filter);
```

## Testing
- **33 comprehensive tests** covering all filter types and edge cases
- **Unit tests** for JSON parsing, type conversion, and error handling
- **Integration tests** with mock language models
- **Edge case testing** for malformed inputs and schema validation
- **Enum support testing** with automatic value discovery

## Documentation
- **Complete API documentation** with Javadoc for all public classes and methods
- **Integration guide** in LangChain4j docs structure (`docs/integrations/filter-builders/`)
- **Comprehensive examples** including real-world RAG application scenarios
- **Module README** with quick start guide and best practices

### General
- [x] No breaking changes introduced
- [x] Unit and integration tests added (33 tests total)
- [x] Both positive and negative test cases covered
- [x] Documentation updated with examples and integration guide
- [x] Comprehensive examples provided for real-world usage

### New Maven Module
- [x] Module added to root `pom.xml` in embedding-store-filter-parsers section
- [x] Module added to `langchain4j-bom/pom.xml` for dependency management
- [x] Follows LangChain4j naming conventions (`langchain4j-embedding-store-filter-builder-language`)
- [x] Proper parent POM configuration with correct version references
- [x] Minimal dependencies (only core LangChain4j, Jackson, test dependencies)

### Code Quality
- [x] Java 17 compatibility confirmed
- [x] No Lombok usage (following LangChain4j guidelines)
- [x] Comprehensive Javadoc for all public APIs
- [x] Follows existing code patterns and conventions
- [x] Proper error handling and validation

### Integration
- [x] Compatible with existing LangChain4j Filter ecosystem
- [x] Works with all embedding stores that support filtering
- [x] Supports all existing LangChain4j filter types
- [x] No conflicts with existing modules

## Language Model Compatibility
Tested and compatible with:
- OpenAI GPT-3.5/GPT-4 (via `langchain4j-open-ai`)
- Azure OpenAI
- Other models supporting ResponseFormat.JSON and structured outputs

## Future Enhancements
- Support for more complex logical expressions
- Additional date parsing formats
- Custom filter type extensions
- Performance optimizations for high-volume usage

This module significantly enhances LangChain4j's capabilities for real-world RAG applications by bridging the gap between natural language user input and structured metadata filtering.

### CODE REVIEW

File: docs/docs/integrations/filter-builders/language-model.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+This is particularly useful for implementing hybrid search where users can express both semantic intent and metadata constraints in natural language.
+
+## Maven Dependency
+
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embedding-store-filter-builder-language</artifactId>
+    <version>1.1.0-beta7-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline docs/docs/integrations/filter-builders/language-model.md:21]:
```suggestion
    <version>1.10.0-beta18</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- **Date Intelligence**: Smart handling of relative dates ("last week", "this month")
+- **JSON Schema Validation**: Uses structured outputs for reliable filter generation
+
+## Quick Start
+
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embedding-store-filter-builder-language</artifactId>
+    <version>1.1.0-beta7-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/README.md:29]:
```suggestion
    <version>1.10.0-beta18</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>1.1.0-beta7-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:8]:
```suggestion
        <version>1.10.0-beta18-SNAPSHOT</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    <artifactId>langchain4j-embedding-store-filter-builder-language</artifactId>
+    <name>LangChain4j :: Embedding Store Filter Builder :: Language Model</name>
+    <description>A filter builder that uses language models to convert natural language queries into LangChain4j Filter objects</description>
+
+    <dependencies>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:21]:
```suggestion
            <version>1.10.0-beta18</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:27]:
```suggestion
            <version>1.10.0-beta18</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
+        </dependency>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:26]: is this dependency required?

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-junit-jupiter</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-open-ai</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:70]:
```suggestion
            <version>1.10.0-SNAPSHOT</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/pom.xml:77]:
```suggestion
            <version>1.10.0-SNAPSHOT</version>
```

File: langchain4j-bom/pom.xml [hunk 9]
```diff
@@ -538,6 +538,12 @@
                 <version>${langchain4j.beta.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>dev.langchain4j</groupId>
+                <artifactId>langchain4j-embedding-store-filter-builder-language</artifactId>
+                <version>1.1.0-beta7-SNAPSHOT</version>
```
Comment (dliubarskyi) [inline langchain4j-bom/pom.xml:544]:
```suggestion
                <version>1.10.0-beta18-SNAPSHOT</version>
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/LanguageModelFilterBuilderExample.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+import dev.langchain4j.store.embedding.filter.Filter;
+
+import java.time.LocalDate;
+import java.util.List;
+
+/**
+ * Comprehensive example demonstrating the Language Model Filter Builder
+ * for converting natural language queries into structured filters.
+ */
+public class LanguageModelFilterBuilderExample {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/LanguageModelFilterBuilderExample.java:23]: Please remove this example from the src/main

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/EmbeddingStoreContentRetrieverDelegate.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+ * A ContentRetriever delegate that wraps an EmbeddingStoreContentRetriever
+ * and adds intelligent query parsing capabilities using LanguageModelFilterBuilder.
+ * 
+ * This delegate:
+ * 1. Parses incoming queries to extract filters and limits
+ * 2. Creates a clean query for embedding generation
+ * 3. Uses the existing dynamicFilter and dynamicMaxResults mechanisms
+ * 4. Preserves all other EmbeddingStoreContentRetriever functionality
+ */
+public class EmbeddingStoreContentRetrieverDelegate implements ContentRetriever {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/EmbeddingStoreContentRetrieverDelegate.java:26]: WOuld be nice to have a more descriptive name, perhaps something containing "LanguageModelFilter"

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/EmbeddingStoreContentRetrieverDelegateExample.java [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Complete working example demonstrating the EmbeddingStoreContentRetrieverDelegate for hybrid search.
+ * This example creates an in-memory embedding store with sample data and shows real hybrid search results.
+ */
+public class EmbeddingStoreContentRetrieverDelegateExample {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/EmbeddingStoreContentRetrieverDelegateExample.java:22]: please remove all examples from src/main

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/FilterResult.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    /**
+     * Gets the structured filter to apply to metadata fields.
+     * This filter constrains the search to documents matching specific criteria
+     * (e.g., publication date, author, category, etc.).
+     * 
+     * @return the filter, or null if no structured filtering is needed
+     */
+    public Filter getFilter() {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/FilterResult.java:27]:
```suggestion
    public Filter filter() {
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/FilterResult.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    /**
+     * Gets the modified query for semantic/embedding-based search.
+     * This query focuses on the semantic content and removes structural constraints
+     * that are better handled by filters.
+     * 
+     * @return the modified query for embedding search
+     */
+    public String getModifiedQuery() {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/FilterResult.java:38]:
```suggestion
    public String modifiedQuery() {
```

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/LanguageModelJsonFilterBuilder.java [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+ * 
+ * @since 1.0.0
+ */
+public class LanguageModelJsonFilterBuilder {
+
+    private final ChatModel chatModel;
+    private final TableDefinition tableDefinition;
+    private final ObjectMapper objectMapper;
+
+    private static final PromptTemplate FILTER_PROMPT = PromptTemplate.from("""
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/LanguageModelJsonFilterBuilder.java:64]: Please allow users to override this default prompt

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/TableDefinition.java [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+    public TableDefinition(List<ColumnDefinition> columns) {
+        this.columns = Objects.requireNonNull(columns, "columns cannot be null");
+    }
+
+    /**
+     * Returns the list of column definitions for this table.
+     * 
+     * @return an unmodifiable list of column definitions
+     */
+    public List<ColumnDefinition> getColumns() {
```
Comment (dliubarskyi) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-builder-language/src/main/java/dev/langchain4j/store/embedding/filter/builder/language/TableDefinition.java:55]:
```suggestion
    public List<ColumnDefinition> columns() {
```
Please check everywhere that there is no "get" in getter names

**Review by dliubarskyi — COMMENTED** [review]
@andyflury thank you and sorry for the delay!
